File: tools/analyze_jsx.js
const fs = require('fs');
const path = 'd:/Project WebApp/axon-ecosystem/frontend/src/app/dashboard/attendance/history/page.tsx';
const text = fs.readFileSync(path,'utf8');
const pattern = /<(/?)([A-Za-z0-9_.:-]+)([^>]*)>/g;
let match;
const stack = [];
const lineStarts = [];
for (let i=0;i<text.length;i++) if (text[i]=='\n') lineStarts.push(i+1);
lineStarts.unshift(0);
function getLine(pos){
  for(let i=0;i<lineStarts.length;i++){
    if (i+1===lineStarts.length || lineStarts[i+1]>pos) return i+1;
  }
  return lineStarts.length;
}
while((match=pattern.exec(text))){
  const start = match.index;
  const line = getLine(start);
  const closing = match[1]==='/'
  const tag = match[2];
  const rest = match[3];
  const selfClosing = /\/$/.test(rest.trim()) || /\/>$/.test(rest.trim());
  if(['br','img','input'].includes(tag)) continue;
  if(closing){
    if(stack.length && stack[stack.length-1].tag===tag){ stack.pop(); }
    else console.log('Unmatched closing',tag,'line',line,'top', stack.length?stack[stack.length-1]:null);
  } else {
    if(!selfClosing){ stack.push({tag,line}); }
  }
}
console.log('Unclosed tags tail:');
console.log(stack.slice(-30));
if(stack.length) console.log('First unclosed:',stack[0]);
else console.log('No unclosed tags found.');

File: tools/analyze_jsx.py
from pathlib import Path
import re
p=Path(r'D:\Project WebApp\axon-ecosystem\frontend\src\app\dashboard\attendance\history\page.tsx')
text=p.read_text(encoding='utf-8')
# simple tag regex
pattern=re.compile(r'<(/?)([A-Za-z0-9_.:-]+)([^>]*)>')
stack=[]
line_starts=[0]
for i,ch in enumerate(text):
    if ch=='\n': line_starts.append(i+1)

for m in pattern.finditer(text):
    start=m.start()
    line = next((i for i,s in enumerate(line_starts) if s>start), len(line_starts))-1
    closing = m.group(1)=='/'
    tag = m.group(2)
    rest = m.group(3)
    self_closing = rest.strip().endswith('/') or rest.strip().endswith('/>')
    # ignore fragments (React fragments) and expressions
    if tag.lower() in ['br','img','input']:
        continue
    if closing:
        if stack and stack[-1][0]==tag:
            stack.pop()
        else:
            print('Unmatched closing',tag,'at line',line+1,'top',stack[-1] if stack else None)
    else:
        if not self_closing:
            stack.append((tag,line+1))

print('Stack tail (unclosed tags):')
for item in stack[-20:]:
    print(item)

a=stack
if a:
    print('\nFirst unclosed tag:',a[0])
else:
    print('No unclosed tags found')
